File: 2016/8.jl
include("utils.jl")
using .AoCUtils

function printScreen(screen)
    s = size(screen)
    for y in 1:s[2]
        for x in 1:s[1]
            if screen[x, y] == 1
                print("#")
            else
                print(".")
            end
        end
        println("")
    end
end

function display(lines, x, y)
    screen = zeros(Int8, x, y)

    for l in lines
        if startswith(l, "rect")
            args = split(l, " ")[2]
            A, B = parse.(Int, split(args, "x"))
            screen[1:A, 1:B] .= 1
        elseif startswith(l, "rotate")
            args = split(l, " ")
            if args[2] == "row"
                row = parse(Int, split(args[3], "y=")[2]) + 1
                by = parse(Int, args[end])
                screen[:, row] = circshift(screen[:, row], by)
            elseif args[2] == "column"
                column = parse(Int, split(args[3], "x=")[2]) + 1
                by = parse(Int, args[end])
                screen[column, :] = circshift(screen[column, :], by)
            end
        end
    end
    printScreen(screen)
    println()
    println("🎅 $(sum(screen))")
end

lines = readLines("input_8.txt")
display(lines, 50, 6)

File: 2016/8.py
from utils import readInput, printMap
import numpy as np

display = np.zeros(shape=(50, 6))
lines = readInput("input_8.txt")
for l in lines:
    if l.startswith("rect"):
        x, y = l.split(" ")[1].split("x")
        x = int(x)
        y = int(y)
        display[:x, :y] = 1
    elif l.startswith("rotate"):
        items = l.split(" ")
        if items[1] == "column":
            col = int(items[2][2:])
            amount = int(items[-1].strip())
            display[col, :] = np.roll(display[col, :], amount)
        elif items[1] == "row":
            row = int(items[2][2:])
            amount = int(items[-1].strip())
            display[:, row] = np.roll(display[:, row], amount)


print ("🎄Part 1: {}".format((display==1).sum()))
print("🎁🎄Part 2: ")
printMap(display)
